File: consorcio/lib/domain/services/supply_folder_service.dart
import '../entities/image_folder.dart';
import '../entities/supply.dart';
import '../value_objects/geo_location.dart';

const virtualSupplyFolderPrefix = 'virtual:';

String supplyFolderDocId(String areaId, String routeCode) {
  return 'sf_${areaId}_$routeCode';
}

final _supplyFolderIdRe = RegExp(
  r'^sf_([0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12})_(\d{7,12})$',
);

({String areaId, String routeCode})? parseSupplyFolderDocId(String id) {
  final match = _supplyFolderIdRe.firstMatch(id);
  if (match == null) return null;
  return (areaId: match.group(1)!, routeCode: match.group(2)!);
}

String virtualSupplyFolderId(String routeCode) {
  return '$virtualSupplyFolderPrefix$routeCode';
}

bool isVirtualSupplyFolderId(String id) {
  return id.startsWith(virtualSupplyFolderPrefix);
}

ImageFolder folderFromSupply({
  required String areaId,
  required String areaName,
  required Supply supply,
  ImageFolder? existing,
}) {
  if (existing != null) return existing;
  final now = DateTime.now();
  return ImageFolder(
    id: supplyFolderDocId(areaId, supply.routeCode),
    areaId: areaId,
    areaName: areaName,
    name: supply.routeCode,
    description: 'Suministro',
    ownerId: '',
    ownerName: '',
    assignToAllTechnicians: true,
    assignedTechnicianIds: const [],
    assignedTechnicianNames: const [],
    imageCount: 0,
    createdAt: now,
    updatedAt: now,
    routeCode: supply.routeCode,
    location: GeoLocation(
      latitude: supply.latitude,
      longitude: supply.longitude,
    ),
  );
}

File: src/domain/services/SupplyFolderService.ts
import type { ImageFolder } from '@/domain/entities/ImageFolder'
import type { Supply } from '@/domain/entities/Supply'
import { supplyHasLocation } from '@/domain/entities/Supply'

export const VIRTUAL_SUPPLY_FOLDER_PREFIX = 'virtual:'

export function supplyFolderDocId(areaId: string, routeCode: string): string {
  return `sf_${areaId}_${routeCode}`
}

const SUPPLY_FOLDER_ID_RE =
  /^sf_([0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12})_(\d{7,12})$/

export function parseSupplyFolderDocId(
  id: string,
): { areaId: string; routeCode: string } | null {
  const match = SUPPLY_FOLDER_ID_RE.exec(id)
  if (!match) return null
  return { areaId: match[1], routeCode: match[2] }
}

export function virtualSupplyFolderId(routeCode: string): string {
  return `${VIRTUAL_SUPPLY_FOLDER_PREFIX}${routeCode}`
}

export function isVirtualSupplyFolderId(id: string): boolean {
  return id.startsWith(VIRTUAL_SUPPLY_FOLDER_PREFIX)
}

export function isSupplyFolder(
  folder: Pick<ImageFolder, 'routeCode'>,
): boolean {
  return Boolean(folder.routeCode)
}

export function folderFromSupply(input: {
  areaId: string
  areaName: string
  supply: Supply
  existing?: ImageFolder
}): ImageFolder {
  if (input.existing) return input.existing
  const now = input.supply.updatedAt
  return {
    id: supplyFolderDocId(input.areaId, input.supply.routeCode),
    areaId: input.areaId,
    areaName: input.areaName,
    name: input.supply.routeCode,
    description: 'Suministro',
    ownerId: '',
    ownerName: '',
    assignToAllTechnicians: true,
    assignedTechnicianIds: [],
    assignedTechnicianNames: [],
    imageCount: 0,
    routeCode: input.supply.routeCode,
    location: supplyHasLocation(input.supply)
      ? {
          latitude: input.supply.latitude,
          longitude: input.supply.longitude,
        }
      : undefined,
    createdAt: now,
    updatedAt: now,
  }
}
